### PR TITLE
[2.0] Modularize repo and build time speedup

### PR DESCRIPTION
**Pull request by cromefire**
- [x] Speed up builds
- [x] Adding lerna
- [x] Extracting packages

### CODE REVIEW

**Comment by cromefire** [comment]
@prateekbh I think we should consider adding https://github.com/wip/app to not accidentally merge WIP PRs. I see me pressing that big green button by accident...

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/df213f456fd482463a503aeb6a6c1d88/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441013772](https://travis-ci.org/prateekbh/preact-material-components/builds/441013772), for commit [01ce16f](https://github.com/prateekbh/preact-material-components/pull/1090/commits/01ce16fba264b5edb42707d7f0c0e10059fdd8de).

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/bff46890e6b10fd63583b49c4406c352/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441015887](https://travis-ci.org/prateekbh/preact-material-components/builds/441015887), for commit [279b61d](https://github.com/prateekbh/preact-material-components/pull/1090/commits/279b61d63fc267e2f952b5d8e1f7f0973605c508).

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/b66105563b1a294723f5ba4852d51cfa/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441142868](https://travis-ci.org/prateekbh/preact-material-components/builds/441142868), for commit [10884d8](https://github.com/prateekbh/preact-material-components/pull/1090/commits/10884d877e4a2ceda00691e859b37555e6270030).

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/af5e1ed2d65d35010d77a55b543bd616/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441143324](https://travis-ci.org/prateekbh/preact-material-components/builds/441143324), for commit [79adf2a](https://github.com/prateekbh/preact-material-components/pull/1090/commits/79adf2a76ac6b3650d7741fa235c74f4e001cf20).

**Comment by artipost** [comment]
Failed pictures (25): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/e91383867b0af301ab7aff2c9eed341d/raw/64facbacf59b23c7721d03e66c8ed41276f44a2a/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441271964](https://travis-ci.org/prateekbh/preact-material-components/builds/441271964), for commit [2315176](https://github.com/prateekbh/preact-material-components/pull/1090/commits/2315176ed0a29b07e94335105eb5fb888d51b9b4).

**Comment by cromefire** [comment]
@prateekbh right now I've shortened the build time from 5:41 run and 6:41 total to 3:07 run and 4:04 total.

Things I've changed:
- `yarn` instead of `npm`
- Ubuntu 14.04 instead of OSX
- `foxr` and Firefox instead of `puppeteer` and Chrome
- `http` and `serve-static` instead of `serve`

I hope you are okay with that

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/2e429b23c9f1dc6f7afbc3bab0724fae/raw/41a7ce5f21a72524b008a6ed0c5464ef4e030a81/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441374864](https://travis-ci.org/prateekbh/preact-material-components/builds/441374864), for commit [d6a9bb0](https://github.com/prateekbh/preact-material-components/pull/1090/commits/d6a9bb0f922ea8602babd56b890b03259cc49030).

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/e0149913e5703e8117b8d0c5315d6676/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441379758](https://travis-ci.org/prateekbh/preact-material-components/builds/441379758), for commit [4f4348d](https://github.com/prateekbh/preact-material-components/pull/1090/commits/4f4348d7ddccf3619c809daa7f6a3024398d42ef).

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/4f109b91fef64c4fa9ecc32afee76da6/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [441380443](https://travis-ci.org/prateekbh/preact-material-components/builds/441380443), for commit [d8f5de1](https://github.com/prateekbh/preact-material-components/pull/1090/commits/d8f5de1e8fe855ce1c2cf573b8b512b8fc0e3aa3).

**Comment by prateekbh** [comment]
@cromefire I am good with the changes except that I am a little skeptical about removing puppeteer.

**Comment by cromefire** [comment]
Right now we are running faster even with `narrow` enabled, which got never implemented.

`foxr`s API is quite similar to `puppeteer`.

We could save another few seconds, when the animations wouln't be there, but right now, we have to wait on their completion.

**Comment by cromefire** [comment]
If you whant to see it in action grab https://github.com/prateekbh/preact-material-components/pull/1090/commits/2a8491e37b1db38af5443fc6254c2bd15cf5d550 and run 

```
yarn install
yarn build
cd docs
yarn install
SHOW_FIREFOX=true yarn test
```
It will open Firefox in not headless mode so you can see, what it does (You need Firefox installed and images produced in not headless mode are invalid).

**Comment by cromefire** [comment]
15 down, 15 to go, I'm half way there, just porting and splitting the components

**Comment by cromefire** [comment]
@prateekbh I think there is nothing against bundling the 4 `Tab*` components into one package, is there?

**Comment by prateekbh** [comment]
@cromefire I guess that'd be fine

**Comment by prateekbh** [comment]
@cromefire can we merge this as is?
So that I have some base to help and pitch in the work... We'll fix tests as we go

**Comment by cromefire** [comment]
No I'm still working on it I should have more time in the next days. The tests should be fixed automatically once I migrated everything and fixed the links for the docs.

**Comment by cromefire** [comment]
~ 3 hours of work left

**Comment by prateekbh** [comment]
Gotcha, thanks for this gigantic effort do take your time

**Comment by prateekbh** [comment]
we have @preact-material as well.

**Comment by cromefire** [comment]
Might be nice

**Comment by digitalica** [comment]
looking good! nice work Chromefire.

**Comment by prateekbh** [comment]
is there any way i can start to pitch in?

**Comment by cromefire** [comment]
You are the `preact-cli` expert, so if you please could have a look at the docs, I got stuck there

**Comment by prateekbh** [comment]
where did the tests folder go?

**Comment by prateekbh** [comment]
@cromefire there is a tslintfix in this branch and then there is prettier fix.
We have a bot testing for prettier changes, lets just use that and not both

**Comment by prateekbh** [comment]
Also why did we movedd from pupeteer to foxr?

**Comment by cromefire** [comment]
> there is a tslintfix in this branch and then there is prettier fix.
We have a bot testing for prettier changes, lets just use that and not both

Prettier can't lint all the things, because it's mainly for javascript, but we should merge both into one cmd

> Also why did we movedd from pupeteer to foxr?

Because this shortens the build drastically, in the long run I want to use `selenium-webdriver` that can use both firefox and chrome, but there is currently a problem I'm waiting to be fixed

**Comment by prateekbh** [comment]
puppeteer has support both chrome and firefox now as well.(experimental though but still actively being worked upon)

**Comment by cromefire** [comment]
> puppeteer has support both chrome and firefox now as well.(experimental though but still actively being worked upon)

Great I didn't know (and firefox is much faster in travis) let's get this fixed, merge it into `2.0` and I will switch back to `puppeteer`

**Comment by prateekbh** [comment]
@cromefire i am bit skeptical about the change in pkg name.. this'll be bit awkward can we stick to `preact-material-components` itself?

**Comment by cromefire** [comment]
Sure, just search & replace

**Comment by prateekbh** [comment]
i am publishing 2.0.0-aplha1 to test docs

**Comment by prateekbh** [comment]
🤦🏻‍♂️sorry

**Comment by cromefire** [comment]
Where you able to figure out, why the docs are failing to build?

**Comment by prateekbh** [comment]
still working on it... the `sass-loader` is ugly, more uglier with multiple `node_modules` folder

**Comment by prateekbh** [comment]
tests running again but artipost not reporting them back..
can i summon @filoozom ?

**Comment by filoozom** [comment]
@prateekbh You sure can 😉 Everything seems to be working on my end though, did anything change in the testing scripts?

**Comment by prateekbh** [comment]
yeah we split it into three different tasks... @cromefire is running this release i am merely helping

**Comment by filoozom** [comment]
Yeah I saw it, looking at the scripts @artipost still seems to be there, but I don't see logs indicating that it is being executed (no "archive is done", "uploading" and all that).

File: scripts/travis.sh [hunk 1]
```diff
@@ -1,60 +1,27 @@
...
-# always run tests for docs
-if [ $1 == "docs" ]
+# Tests for docs
+if [[ "${1}" == "docs" ]]
 then
-    test
+    set -ev
+    yarn build
+    cd packages/docs
+    yarn test
```
Comment (filoozom) [inline scripts/travis.sh]: Looks like you should add something like this back:
```
    if npm run | grep test:travis > /dev/null
    then
        npm run test:travis
    else
        npm test
    fi
```

Basically you're not even calling the @artipost thing right now 😉

Comment (cromefire) [reply scripts/travis.sh]: Fixed

> (no "archive is done", "uploading" and all that)

The doc build crashes before anyway right now

**Comment by cromefire** [comment]
A good thing anyone can halp with is reviewing and checking for consistency

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/eda88fa69fe86c1fa2a7ffed5579bf6d/raw/6b95adcd27b6311d6dfe2eedd342de99531fb696/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [480227013](https://travis-ci.org/prateekbh/preact-material-components/builds/480227013), for commit [539628d](https://github.com/prateekbh/preact-material-components/pull/1090/commits/539628de44e9be2a43a16ae8e889a7c95d18cb63).

**Comment by cromefire** [comment]
Please revert this one: https://github.com/prateekbh/preact-material-components/blob/6530545b3c126bfa08d0bd53587c99c6da2b27d7/packages/base/package.json it had a sense, because the versions where all set to `0.39.0` but some packages have up to `0.39.3`

**Comment by cromefire** [comment]
Also after this initial PR is done, I thought anyone can pick some components individually and improve them seperatly

**Comment by cromefire** [comment]
So that we could have like 1 PR / Component

**Comment by cromefire** [comment]
@prateekbh Please fix `/packages/docs/tests/src/screenshot.js` for puppeteer, I'm not available for a few days

**Comment by prateekbh** [comment]
@cromefire there's actually nothing to fix, its just not able to find the screenshot to compare.
I'll add them in and we should be good.

Also I feel that we should merge this once the tests pass and then pick components one by one -> upgrade them -> add any gestures work they need -> release

**Comment by cromefire** [comment]
> Also I feel that we should merge this once the tests pass and then pick components one by one -> upgrade them -> add any gestures work they need -> release

This exactly what I proposed in a previous comment

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/fbd9f070e5428541b250e3209512f581/raw/8839e28872b0dab8c8dcdac12f3dcc25d2621248/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [482985341](https://travis-ci.org/prateekbh/preact-material-components/builds/482985341), for commit [7e6581e](https://github.com/prateekbh/preact-material-components/pull/1090/commits/7e6581eae5d46d7dbc6d02feefdbe0388db01c1a).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/a5d5b378154214756f04b7ef5d628339/raw/1e0d3ef434fdb90b7348ba39ecba34d3512238b2/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [482989583](https://travis-ci.org/prateekbh/preact-material-components/builds/482989583), for commit [039f8df](https://github.com/prateekbh/preact-material-components/pull/1090/commits/039f8df5f258c4208e85edd419db2dd92e69b4c1).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/38949e34962037cca26e9f214cb7c274/raw/7de5e493067147681cc28dc44e75cfb6a1184c8a/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [483013265](https://travis-ci.org/prateekbh/preact-material-components/builds/483013265), for commit [277563b](https://github.com/prateekbh/preact-material-components/pull/1090/commits/277563bbd70220ef2490675d311a631bba5aa8b9).

**Comment by cromefire** [comment]
Are you sure that import was wrong? There are both `action` and `actions` I think, and my IDE would have warned me

**Comment by prateekbh** [comment]
well i'll check it again once i fix the screenshot problem

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/d5a61e507bd94f9a36ce1a7ca5aa7da8/raw/386df3acf66ba64c51a76f596730d4ec420d850c/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [483027837](https://travis-ci.org/prateekbh/preact-material-components/builds/483027837), for commit [3adb84c](https://github.com/prateekbh/preact-material-components/pull/1090/commits/3adb84c383240eda4e9c7af8ca9940bfc0caa833).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/11d7258d307bc320168c92daf7eff499/raw/f82e525f842eb86e2a80f1f323846f5f3b705851/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [483048689](https://travis-ci.org/prateekbh/preact-material-components/builds/483048689), for commit [584f83c](https://github.com/prateekbh/preact-material-components/pull/1090/commits/584f83cdaa3c3f3c583143dac7e2da28648b3e6c).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/e2d2b74d7a1b5c69e3c9c1ebbc188876/raw/fa63ff3d9e37a76c33a2083a3f01a614ff3b64b0/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [483060449](https://travis-ci.org/prateekbh/preact-material-components/builds/483060449), for commit [132402f](https://github.com/prateekbh/preact-material-components/pull/1090/commits/132402fdb77b74f1c5c25e248eeb6354e66fc57a).

**Comment by cromefire** [comment]
Did you move the docs out of packages (stated in a commit message)? That would break all the things.
Could you describe more precisely what went wrong? I've not much time until Saturday.

**Comment by cromefire** [comment]
Forget that moving docs thing I saw you moved them back

**Comment by prateekbh** [comment]
@cromefire i dont know what's causing the sass error, can we just use css files for now?

also we'll need the script to copy them over like before

**Comment by cromefire** [comment]
We can import the css files from sass right?

**Comment by cromefire** [comment]
Do you know, what that `proxy-loader` is and whan it is used by `preact-cli`?

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/f0c4d2e420e05b07309533dea2ea66f5/raw/8823ecf5ad27e7595be926bf47ed1701d5e64690/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484384138](https://travis-ci.org/prateekbh/preact-material-components/builds/484384138), for commit [394168f](https://github.com/prateekbh/preact-material-components/pull/1090/commits/394168f8d04733d54b3b4ff1ef5c787bef8ce8d9).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip]

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484384392](https://travis-ci.org/prateekbh/preact-material-components/builds/484384392), for commit [eaf6e2b](https://github.com/prateekbh/preact-material-components/pull/1090/commits/eaf6e2b2e0be4316ece4b9a3a06a16d913afdb2c).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/5e0481fe9af0f5cb8010a85d2eabc833/raw/068a369efccb46eee05e011a4cb91bc6176a4827/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484412371](https://travis-ci.org/prateekbh/preact-material-components/builds/484412371), for commit [5e77da9](https://github.com/prateekbh/preact-material-components/pull/1090/commits/5e77da940535c3c5f5d23026fa26182f54d926e4).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/4fc00a1736e346761af894d25b3c8b3b/raw/3f6735b4286ee0247d654763581172d96f9ff5f1/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484443924](https://travis-ci.org/prateekbh/preact-material-components/builds/484443924), for commit [770c50f](https://github.com/prateekbh/preact-material-components/pull/1090/commits/770c50f4ba0ac0ea13982fa5eb14f8d2136e5aef).

**Comment by cromefire** [comment]
Got everything fixed

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip]

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484469382](https://travis-ci.org/prateekbh/preact-material-components/builds/484469382), for commit [fabd799](https://github.com/prateekbh/preact-material-components/pull/1090/commits/fabd799ad547f9a87b171368a09e6f0965d88976).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/a988e29438a7cd59310f9f18227d2ea4/raw/c603240b7fd1c3132fc13468d7f711313312d640/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484469382](https://travis-ci.org/prateekbh/preact-material-components/builds/484469382), for commit [fabd799](https://github.com/prateekbh/preact-material-components/pull/1090/commits/fabd799ad547f9a87b171368a09e6f0965d88976).

**Comment by cromefire** [comment]
Oh and I speeded up the build again

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/1d4a7f9a7b39c5a0c755a6b2d016a95a/raw/29c52fc06f14d4ef8e8c7fb0f8411bb93b9d9f48/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484489260](https://travis-ci.org/prateekbh/preact-material-components/builds/484489260), for commit [23acd38](https://github.com/prateekbh/preact-material-components/pull/1090/commits/23acd38db0aeaf16d468c0b21f3cfc80f9518ccc).

**Comment by cromefire** [comment]
> i dont even know what that means... but that does seem to fix the problem

@prateekbh it means that webpack will exclude things that you did you didn't import, if you are using es6 imports

**Comment by prateekbh** [comment]
@cromefire i guess the parallelization broke the tests... just checking.

I suspect this cuz some tests pass randomly

**Comment by artipost** [comment]
Failed pictures (7): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/4590ecb9118afc4f5b9b35060136f7ae/raw/830d840b2ea3eb7268290c20b2ec65b7c2ac103a/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484615984](https://travis-ci.org/prateekbh/preact-material-components/builds/484615984), for commit [9771a46](https://github.com/prateekbh/preact-material-components/pull/1090/commits/9771a4631405465ad8e4ecd1480e75ce6a30cd2c).

**Comment by prateekbh** [comment]
@cromefire all fixed lets merge this and then do anything further.. this has become a mammoth PR

**Comment by cromefire** [comment]
Yes let me tweak a bit at the test parallelism (I think I should limit by CPU count) and then we can merge it.

I'm sorry, I needed some time to figure out, what works the best.

**Comment by cromefire** [comment]
If parallel test don't work then I will give it up and merge

**Comment by artipost** [comment]
Failed pictures (25): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/720d3ed182c03c09bd435e35c0da9c82/raw/de50acaf1f5f5167951ef600f083670511c1f306/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484708320](https://travis-ci.org/prateekbh/preact-material-components/builds/484708320), for commit [e7eec26](https://github.com/prateekbh/preact-material-components/pull/1090/commits/e7eec26ef0796d3110d99bfd23d12c11171123c0).

**Comment by artipost** [comment]
Failed pictures (1): [failed-pictures.zip]

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484710344](https://travis-ci.org/prateekbh/preact-material-components/builds/484710344), for commit [624d2b7](https://github.com/prateekbh/preact-material-components/pull/1090/commits/624d2b70cf7d9863491ccd104e2a3d923dc2a815).

**Comment by artipost** [comment]
Failed pictures (1): [failed-pictures.zip]

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484714361](https://travis-ci.org/prateekbh/preact-material-components/builds/484714361), for commit [0ef4526](https://github.com/prateekbh/preact-material-components/pull/1090/commits/0ef45266d11033c7d580c7828a083ca83bd00928).

**Comment by artipost** [comment]
Failed pictures (1): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/35fbc6e7253f125bc21f87f7ac825f1b/raw/bb7548327c678693ee550a660b209a8ed0f75ea4/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [484715855](https://travis-ci.org/prateekbh/preact-material-components/builds/484715855), for commit [cac87f3](https://github.com/prateekbh/preact-material-components/pull/1090/commits/cac87f3019512a831ce3eace6e64818acd3a0f47).

**Pull request merged**